### PR TITLE
Fix FileUtilsInstrumentationTest crash when SD card is unavailable

### LINKED ISSUE #4761: FileUtilsInstrumentationTest fails with ArrayIndexOutOfBoundsException on devices without SD card
**Describe the bug**
The instrumentation test testGetLocalFilePathByUri in FileUtilsInstrumentationTest crashes with an ArrayIndexOutOfBoundsException when running on devices or emulators that do not have external storage (SD card).

The test assumes that context.getExternalFilesDirs("") always returns at least two directories (internal storage and SD card). However, on many devices only internal storage is available, so the array contains only one element. Accessing index 1 results in a crash.

**Expected behavior**
The test should not crash when external storage is unavailable.

**Steps to reproduce the behavior:**
1. Run FileUtilsInstrumentationTest
2.Run on an emulator or device without SD card
3.The test fails with ArrayIndexOutOfBoundsException

**Screenshots**
Image

**Environment**
- Version of Kiwix Android : Latest main branch
- Device : Android Emulator(OPPO CPH2495)
- OS version : Android 14

**Logs**
java.lang.ArrayIndexOutOfBoundsException: length=1; index=1 at FileUtilsInstrumentationTest.kt:426

The issue appears to be caused by accessing getExternalFilesDirs("")[1] without verifying that the array contains more than one element.

### PR DESCRIPTION
**Pull request by MohanReddy369**
Fixes #4761

**Solution:**  
Added a null and size check before accessing the second element of `getExternalFilesDirs("")`. If no SD card is available, `sdCardPath` will be `null`, preventing the crash.

**Changes Made:**  
- Checked if `externalDirs` contains more than one element before accessing index 1.  
- Safely assigned `sdCardPath` to `null` if no SD card directory exists.  

**Screenshots** 
testGetLocalFilePathByUri

### CODE REVIEW

**Comment by harshsomankar123-tech** [comment]
Hi @MohanReddy369, thanks for working on this!

I had already worked on a similar fix in PR #4618,where I used`getOrNull(1)` on `getExternalFilesDirs("")`, which safely returns `null` when the index doesn't exist  preventing the `IndexOutOfBoundsException` on devices/emulators without an SD card. I didn’t see your issue/PR earlier appreciate you looking into it.

While testing my PR, I also found a few related issues :
• Replaced `CoroutineScope(Dispatchers.Main).launch` with `runBlocking` so assertion failures are properly detected.
• Corrected the `getDocumentId` mock to use `uri.pathSegments[1]` for external storage URIs.
• Fixed a small edge case in `FileUtils.kt` where `sdCardOrUsbMainPath` could be interpolated as `"null"` instead of returning `null`.
• Made the USB test conditional instead of assuming `/mnt/media_rw/USB/...` on emulators without USB storage.
If the maintainers prefer my approach, you can also refer to the fixes included in your PR.

Since my PR is fairly large and may take some time to review. I can remove or adjust my related changes if needed to avoid conflicts.u can take this issue.

@MohitMaliFtechiz what do you think would be the better approach here?

**Comment by MohanReddy369** [comment]
@harshsomankar123-tech ,Your PR looks good and handles additional edge cases. I’m fine with using your approach if the maintainers agree, otherwise we can merge mine.

File: app/src/androidTest/java/org/kiwix/kiwixmobile/utils/files/FileUtilsInstrumentationTest.kt [hunk 1]
```diff
@@ -420,10 +420,15 @@ class FileUtilsInstrumentationTest {
...
       }
     }
-    // get the SD card path
+    // Get all external storage directories (internal storage and possible SD card)
+    val externalDirs = context?.getExternalFilesDirs("")
+    // Safely get the SD card path if it exists (avoid crash on devices without SD card)
     val sdCardPath =
-      context?.getExternalFilesDirs("")
-        ?.get(1)?.path?.substringBefore("/Android")
+      if (externalDirs != null && externalDirs.size > 1) {
```
Comment (MohitMaliFtechiz) [inline app/src/androidTest/java/org/kiwix/kiwixmobile/utils/files/FileUtilsInstrumentationTest.kt]: This condition will return `null`, and we are testing a case based on it. So this will be null, and it will not properly test the below edge case. Because if it is `null`, then testing this scenario will give a `false` result(even if the test is passed). Also, it can not be detected(Which fails the purpose of the test).

```kotlin
DummyUrlData(
          null,
          null,
          "$sdCardPath/$commonPath",
          null,
          Uri.parse(
            primaryStorageUriPrefix +
              sdCardPath?.substringAfter("storage/") +
              "%3A$commonUri"
          )
        )
```

Comment (MohanReddy369) [reply app/src/androidTest/java/org/kiwix/kiwixmobile/utils/files/FileUtilsInstrumentationTest.kt]: > This condition will return `null`, and we are testing a case based on it. So this will be null, and it will not properly test the below edge case. Because if it is `null`, then testing this scenario will give a `false` result(even if the test is passed). Also, it can not be detected(Which fails the purpose of the test).
> 
> ```kotlin
> DummyUrlData(
>           null,
>           null,
>           "$sdCardPath/$commonPath",
>           null,
>           Uri.parse(
>             primaryStorageUriPrefix +
>               sdCardPath?.substringAfter("storage/") +
>               "%3A$commonUri"
>           )
>         )
> ```

@MohitMaliFtechiz

Do you mean that if a device does not have an SD card, then `sdCardPath` becomes `null`, but the test still uses `sdCardPath` in `DummyUrlData`, like this:

```
DummyUrlData(
    null,
    null,
    "$sdCardPath/$commonPath",
    null,
    Uri.parse(
        primaryStorageUriPrefix +
        sdCardPath?.substringAfter("storage/") +
        "%3A$commonUri"
    )
)
```

In that case, if `sdCardPath = null`, the SD card path is not a valid path anymore. The test may still pass, but the SD card scenario would not actually be tested.

Instead of letting the test fail with `ArrayIndexOutOfBoundsException: index 1`, would it make sense to fail the test with a clearer message when the SD card is missing? For example:

```
require(externalDirs!!.size > 1) {
    "SD card must be enabled in emulator to run this test"
}
```

This way the test still fails when the SD card is not available, but the failure message clearly explains the reason instead of throwing an `ArrayIndexOutOfBoundsException`.

Comment (MohitMaliFtechiz) [reply app/src/androidTest/java/org/kiwix/kiwixmobile/utils/files/FileUtilsInstrumentationTest.kt]: @MohanReddy369 Yes, we should improve the error message here so that it will be clear to us what is the problem.

Comment (MohanReddy369) [reply app/src/androidTest/java/org/kiwix/kiwixmobile/utils/files/FileUtilsInstrumentationTest.kt]: > @MohanReddy369 Yes, we should improve the error message here so that it will be clear to us what is the problem.

@MohitMaliFtechiz If the SD card is present, the test will run normally. If the SD card is missing, the test will fail with a clear message such as:
IllegalArgumentException: SD card must be enabled in emulator to run this test.
Is this message okay? and the code is 

val externalDirs = context?.getExternalFilesDirs("")
require(externalDirs != null && externalDirs.size > 1) {
    "SD card must be enabled in emulator to run this test"
}
val sdCardPath = externalDirs[1].path.substringBefore("/Android")

no sd card return message

Comment (MohitMaliFtechiz) [reply app/src/androidTest/java/org/kiwix/kiwixmobile/utils/files/FileUtilsInstrumentationTest.kt]: @MohanReddy369 `SD card must be enabled in the emulator/device to run this test`. We should add this message.

Comment (MohitMaliFtechiz) [reply app/src/androidTest/java/org/kiwix/kiwixmobile/utils/files/FileUtilsInstrumentationTest.kt]: Rebase your PR as well.

Comment (MohanReddy369) [reply app/src/androidTest/java/org/kiwix/kiwixmobile/utils/files/FileUtilsInstrumentationTest.kt]: > @MohanReddy369 `SD card must be enabled in the emulator/device to run this test`. We should add this message.

Hi @MohitMaliFtechiz,

I have updated the test to fail with the message
"SD card must be enabled in the emulator/device to run this test."
fie-utils-instrumentation-test kt

**Review by MohitMaliFtechiz — APPROVED** [review]
@MohanReddy369 Thank you. @kelson42 LGTM.
